### ISSUE TITLE
AutoPages are created before Generators run

### ISSUE THREAD

**Issue by ibrado** [sverrirs/jekyll-paginate-v2#59]
Since AutoPages is triggered [right after the site is loaded](https://github.com/sverrirs/jekyll-paginate-v2/blob/ebdd42f13fb285c3d7ca02a0e90d8229983b47ed/lib/jekyll-paginate-v2/autopages/autoPages.rb#L5-L7) (i.e. immediately after Jekyll starts), autopages for new tags and categories of documents/pages created by other `Jekyll::Generators` are not generated.

e.g. `_plugins/unseen_doc_generator.rb`:

```ruby
module Jekyll
  module UnseenDoc
    class Generator < Jekyll::Generator
      def generate(site)
        unseen_doc = Jekyll::Document.new(File.join(site.source, 'unseen-doc.html'),
          { :site => site, :collection => site.posts })

        unseen_doc.data["layout"] = "page"
        unseen_doc.data["title"] = "An unseen document"
        unseen_doc.data["date"] = Time.new

        unseen_doc.content = "This document's tags/category will not be processed by AutoPages<br/>" +
          "My categories: {{ page.categories | join: ' ' }}<br/>\n" +
          "My tags: {{ page.tags | join: ' '}}<br/>\n"

        unseen_doc.data["tags"] = [ "unseen-tag", "another-tag" ]
        unseen_doc.data["categories"] = [ "unseen-category" ]

        site.posts.docs << unseen_doc
      end
    end
  end
end
```

The resulting post will be listed in the homepage and accessible at e.g. `/2017/12/unseen-doc/` (or whatever the `permalink` dictates), but its category and tags are not processed by AutoPages, even though the tags are listed by this sample `./tags.md`:

```html
<ul>
{% for tag in site.tags %}
  <li><a href="{{ '/tags/' | append: tag[0] | relative_url }}">{{tag[0] }}</a></li>
{% endfor %}
</ul>
```

**Comment by sverrirs** [comment]
Hmm, so the fix might be to try a different event hook for Jekyll, e.g. this line? 
https://github.com/sverrirs/jekyll-paginate-v2/blob/master/lib/jekyll-paginate-v2/autopages/autoPages.rb#L7

**Comment by ibrado** [comment]
I tried to use `pre_render` (even prepared a PR before this bugrep), but it had some bad side effects -- although the tags/category folders were generated, the autopages were empty, i.e. `paginator.posts` wasn't returning anything.

I may have missed somethingm though.

**Comment by ibrado** [comment]
Tried it again just now, just changing `post_read` to `pre_render`. I put some debug stuff in the screenshots below (`I'm here!`before the `paginator.posts` loop and `Title:` just inside.) So weird...

`post_read`:

![post_read](https://user-images.githubusercontent.com/3312209/34037203-cd20c532-e1c2-11e7-91eb-54e58f715a5d.png)

`pre_render`:

![pre_render](https://user-images.githubusercontent.com/3312209/34037213-d6d71ba8-e1c2-11e7-9623-3871b70deadf.png)

**Comment by ibrado** [comment]
Revisiting this... Apparently, changing the hook to `pre_render` makes AutoPages run after JPv2's generator itself, thus the missing posts. Will try to dig in some more.

**Comment by ibrado** [comment]
The solution was not to use a hook at all, but to call the code from generate() as a class method of PaginateV2::AutoPages. Have made a PR for your consideration, @sverrirs ... thanks!

**Comment by sverrirs** [comment]
Thanks Alex, I will take a look :)

**Comment by sverrirs** [comment]
Fixed in #69
